source/agora/test/GetEnrollments.d

```diff
@@ -0,0 +1,138 @@
+        {
+            this.enable_catchup = enable_catchup;
+            super(args);
+        }
+
+        ///
+        protected override void catchupTask () nothrow
+        {
+            if (atomicLoad(*this.enable_catchup))
+                super.catchupTask();
+        }
+    }
+
+    static class CustomAPIManager : TestAPIManager
+    {
+        mixin ForwardCtor!();
+
+        public static shared bool enable_catchup = false;
+
+        /// set base class
+        public override void createNewNode (Config conf, string file, int line)
+        {
+            if (this.nodes.length == GenesisValidators - 1)
+                this.addNewNode!CustomValidator(conf, &this.enable_catchup,
+                    file, line);
+            else
+                super.createNewNode(conf, file, line);
+        }
+    }
+
+    TestConf conf = {
+        outsider_validators : 1,
+        recurring_enrollment : false
+    };
+    auto all_validators = GenesisValidators + conf.outsider_validators;
+    auto network = makeTestNetwork!CustomAPIManager(conf);
+    network.start();
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.waitForDiscovery();
+
+    auto bad_node = network.clients[GenesisValidators - 1];
+    auto outsider = network.nodes[GenesisValidators];
+
+    // generate 18 blocks, 2 short of the enrollments expiring.
+    network.generateBlocks(Height(GenesisValidatorCycle - 2));
+
+    // make sure outsiders are up to date
+    network.expectHeight(iota(GenesisValidators, all_validators),
+        Height(GenesisValidatorCycle - 2));
+
+    // the bad validator becomes unresponsive
+    bad_node.filter!(API.postTransaction);
+
+    // Block 19 we add the freeze utxos for set_b validators
+    // prepare frozen outputs for outsider validators to enroll
+    // we add the frozen utxo for the outsider
+    const key = outsider.getPublicKey().key;
+    network.blocks[0].spendable().drop(1).takeExactly(1)
+        .map!(txb => txb
+            .split([key]).sign(OutputType.Freeze))
+            .each!(tx => network.nodes[0].postTransaction(tx));
+
+    auto target_nodes = iota(GenesisValidators - 1).array;
+    target_nodes ~= GenesisValidators;
+    network.generateBlocks(target_nodes, Height(GenesisValidatorCycle - 1));
+ 
+    // Now we enroll all the validators including the outsider
+    auto enroll = network.clients[all_validators - 1].setRecurringEnrollment(true);
+    iota(GenesisValidators)
+        .each!(i => network.enroll(iota(GenesisValidators), i));
+
+    auto gotten_enroll = bad_node.getEnrollment(enroll.utxo_key);
+    assert(gotten_enroll == Enrollment.init);
+
+    bad_node.clearFilter();
+
+    // enable `catchupTask` for the bad validator and
+    // make sure the bad validator is up to date
+    CustomAPIManager.enable_catchup = true;
+    network.expectHeight([GenesisValidators - 1], Height(GenesisValidatorCycle - 1));
+
+    Thread.sleep(conf.node.enrollment_catchup_interval);
```

Comment:
How long will this sleep? I think the default is too long for the test and should be overridden to maybe 500.msecs?

Reply:
Also I think it is better to use retry with timeout rather than sleep.

Reply:
I will try to use retry.

Reply:
Used retry. It's better.